// @bigcommerce/big-design/dist/es/components/Stepper/Stepper.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
var _excluded = ["className", "style", "steps", "currentStep"];
import React, { memo } from 'react';
import { Flex } from '../Flex';
import { Step } from './Step';
export var Stepper = /*#__PURE__*/memo(function (_ref) {
  var className = _ref.className,
      style = _ref.style,
      steps = _ref.steps,
      _ref$currentStep = _ref.currentStep,
      currentStep = _ref$currentStep === void 0 ? 0 : _ref$currentStep,
      props = _objectWithoutProperties(_ref, _excluded);

  return steps.length ? /*#__PURE__*/React.createElement(Flex, _extends({}, props, {
    "aria-valuenow": currentStep + 1,
    "aria-valuemin": 0,
    "aria-valuemax": steps.length,
    "aria-valuetext": "Step ".concat(currentStep + 1, ": ").concat(steps[currentStep]),
    flexDirection: "row",
    flexWrap: "nowrap",
    marginBottom: "xLarge",
    role: "progressbar"
  }), steps.map(function (text, index) {
    return /*#__PURE__*/React.createElement(Step, {
      index: index,
      selectedIndex: currentStep,
      stepperLength: steps.length,
      text: text,
      key: index
    });
  })) : null;
});
Stepper.displayName = 'Stepper';
