// @bigcommerce/big-design/dist/es/components/Worksheet/hooks/useStore/useStore.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

import create from 'zustand';
import createContext from 'zustand/context';
import { deleteCells, getHiddenRows, mergeCells } from '../../utils';

var _createContext = createContext(),
    Provider = _createContext.Provider,
    useStore = _createContext.useStore;

export { Provider, useStore };
export var createStore = function createStore() {
  return create(function (set) {
    return {
      columns: [],
      editedCells: [],
      editingCell: null,
      expandableRows: {},
      hiddenRows: [],
      invalidCells: [],
      openedModal: null,
      rows: [],
      selectedCells: [],
      selectedRows: [],
      tableRef: null,
      addEditedCells: function addEditedCells(cells) {
        return set(function (state) {
          return _objectSpread(_objectSpread({}, state), {}, {
            editedCells: mergeCells(state.editedCells, cells)
          });
        });
      },
      addInvalidCells: function addInvalidCells(cells) {
        return set(function (state) {
          return _objectSpread(_objectSpread({}, state), {}, {
            invalidCells: mergeCells(state.invalidCells, cells)
          });
        });
      },
      removeInvalidCells: function removeInvalidCells(cells) {
        return set(function (state) {
          return _objectSpread(_objectSpread({}, state), {}, {
            invalidCells: deleteCells(state.invalidCells, cells)
          });
        });
      },
      setColumns: function setColumns(columns) {
        return set(function (state) {
          return _objectSpread(_objectSpread({}, state), {}, {
            columns: columns
          });
        });
      },
      setEditingCell: function setEditingCell(cell) {
        return set(function (state) {
          return _objectSpread(_objectSpread({}, state), {}, {
            editingCell: cell
          });
        });
      },
      setExpandableRows: function setExpandableRows(expandableRows) {
        return set(function (state) {
          return _objectSpread(_objectSpread({}, state), {}, {
            expandableRows: expandableRows,
            hiddenRows: getHiddenRows(expandableRows)
          });
        });
      },
      setHiddenRows: function setHiddenRows(hiddenRows) {
        return set(function (state) {
          return _objectSpread(_objectSpread({}, state), {}, {
            hiddenRows: hiddenRows
          });
        });
      },
      setOpenModal: function setOpenModal(value) {
        return set(function (state) {
          return _objectSpread(_objectSpread({}, state), {}, {
            openedModal: value
          });
        });
      },
      setRows: function setRows(rows) {
        return set(function (state) {
          return _objectSpread(_objectSpread({}, state), {}, {
            rows: rows
          });
        });
      },
      setSelectedCells: function setSelectedCells(cells) {
        return set(function (state) {
          return _objectSpread(_objectSpread({}, state), {}, {
            selectedCells: cells
          });
        });
      },
      setSelectedRows: function setSelectedRows(rowIndexes) {
        return set(function (state) {
          return _objectSpread(_objectSpread({}, state), {}, {
            selectedRows: rowIndexes
          });
        });
      },
      setTableRef: function setTableRef(ref) {
        return set(function (state) {
          return _objectSpread(_objectSpread({}, state), {}, {
            tableRef: ref
          });
        });
      }
    };
  });
};
