// @bigcommerce/big-design/dist/es/components/Worksheet/editors/CheckboxEditor/CheckboxEditor.js
import React, { useCallback, useEffect } from 'react';
import { typedMemo } from '../../../../utils';
import { Checkbox } from '../../../Checkbox';
import { CheckboxWrapper } from './styled';

var InternalCheckboxEditor = function InternalCheckboxEditor(_ref) {
  var cell = _ref.cell,
      toggle = _ref.toggle,
      onBlur = _ref.onBlur,
      onChange = _ref.onChange;
  var handleChange = useCallback(function () {
    onChange(!cell.value);
  }, [cell, onChange]);
  useEffect(function () {
    // toggle will only return true when a user has pressed
    // `enter` or `space` when a checkbox cell is selected.
    // It is virtually the same as clicking on the checkbox.
    if (toggle) {
      handleChange();
    }
  }, [cell.value, handleChange, toggle]);
  return /*#__PURE__*/React.createElement(CheckboxWrapper, null, /*#__PURE__*/React.createElement(Checkbox, {
    disabled: cell.disabled,
    checked: cell.value,
    hiddenLabel: true,
    label: cell.value ? 'Checked' : 'Unchecked',
    onBlur: onBlur,
    onChange: handleChange
  }));
};

export var CheckboxEditor = typedMemo(InternalCheckboxEditor);
