// @bigcommerce/big-design/dist/cjs/components/Search/Search.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Search = void 0;

var _bigDesignIcons = require("@bigcommerce/big-design-icons");

var _react = _interopRequireDefault(require("react"));

var _Button = require("../Button");

var _Flex = require("../Flex");

var _Form = require("../Form");

var _Input = require("../Input");

var Search = function Search(_ref) {
  var value = _ref.value,
      onChange = _ref.onChange,
      onSubmit = _ref.onSubmit;

  var handleSubmit = function handleSubmit(event) {
    event.preventDefault();
    event.stopPropagation();

    if (typeof onSubmit === 'function') {
      onSubmit(event);
    }
  };

  return /*#__PURE__*/_react.default.createElement(_Form.Form, {
    onSubmit: handleSubmit,
    fullWidth: true
  }, /*#__PURE__*/_react.default.createElement(_Flex.Flex, {
    alignItems: "center",
    backgroundColor: "white",
    flexDirection: "row",
    paddingBottom: "xxSmall"
  }, /*#__PURE__*/_react.default.createElement(_Flex.FlexItem, {
    flexGrow: 1,
    marginRight: "small"
  }, /*#__PURE__*/_react.default.createElement(_Input.Input, {
    "aria-label": "Search",
    placeholder: "Search",
    type: "search",
    value: value,
    onChange: onChange,
    iconLeft: /*#__PURE__*/_react.default.createElement(_bigDesignIcons.SearchIcon, {
      color: "secondary50"
    })
  })), /*#__PURE__*/_react.default.createElement(_Button.Button, {
    mobileWidth: "auto",
    type: "submit",
    variant: "secondary"
  }, "Search")));
};

exports.Search = Search;
