// @bigcommerce/big-design/dist/cjs/components/Worksheet/hooks/useNavigation/useNavigation.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useNavigation = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = require("react");

var _useStore = require("../useStore");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var useNavigation = function useNavigation(selectedCell) {
  var rows = (0, _useStore.useStore)((0, _react.useMemo)(function () {
    return function (state) {
      return state.rows;
    };
  }, []));
  var columns = (0, _useStore.useStore)((0, _react.useMemo)(function () {
    return function (state) {
      return state.columns;
    };
  }, []));
  var hiddenRows = (0, _useStore.useStore)((0, _react.useMemo)(function () {
    return function (state) {
      return state.hiddenRows;
    };
  }, []));
  var setSelectedCells = (0, _useStore.useStore)(function (state) {
    return state.setSelectedCells;
  });
  var setSelectedRows = (0, _useStore.useStore)(function (state) {
    return state.setSelectedRows;
  });
  var isValidPosition = (0, _react.useCallback)(function (position) {
    var rowLength = rows.length;
    var columnsLength = columns.length; // Check to see if the next indexes fit inside the matrix

    return position.rowIndex >= 0 && position.rowIndex < rowLength && position.columnIndex >= 0 && position.columnIndex < columnsLength;
  }, [columns, rows]);
  var isHidden = (0, _react.useCallback)(function (rowIndex) {
    var row = rows[rowIndex];
    var id = row.id;
    return hiddenRows.includes(id);
  }, [hiddenRows, rows]); // This function will look for the next non hidden offset,
  // depending on the direction of the offset.

  var getNextOffset = (0, _react.useCallback)(function (_ref) {
    var columnIndex = _ref.columnIndex,
        rowIndex = _ref.rowIndex;

    if (rowIndex === 0) {
      if (columnIndex > 0) {
        return {
          columnIndex: ++columnIndex,
          rowIndex: rowIndex
        };
      }

      return {
        columnIndex: --columnIndex,
        rowIndex: rowIndex
      };
    }

    if (rowIndex > 0) {
      return {
        rowIndex: ++rowIndex,
        columnIndex: columnIndex
      };
    }

    return {
      rowIndex: --rowIndex,
      columnIndex: columnIndex
    };
  }, []);
  var navigate = (0, _react.useCallback)(function (offset) {
    if (!selectedCell) {
      return;
    }

    var newPosition = {
      columnIndex: selectedCell.columnIndex + offset.columnIndex,
      rowIndex: selectedCell.rowIndex + offset.rowIndex
    };

    if (isValidPosition(newPosition)) {
      if (isHidden(newPosition.rowIndex)) {
        return navigate(getNextOffset(offset));
      }

      var hash = columns[newPosition.columnIndex].hash;
      var type = columns[newPosition.columnIndex].type || 'text';
      var value = rows[newPosition.rowIndex][hash];
      var disabled = columns[newPosition.columnIndex].disabled || false;

      var cell = _objectSpread(_objectSpread({}, newPosition), {}, {
        disabled: disabled,
        hash: hash,
        type: type,
        value: value
      });

      setSelectedCells([cell]);
      setSelectedRows([newPosition.rowIndex]);
    }
  }, [columns, getNextOffset, isHidden, isValidPosition, rows, selectedCell, setSelectedCells, setSelectedRows]);
  return (0, _react.useMemo)(function () {
    return {
      navigate: navigate
    };
  }, [navigate]);
};

exports.useNavigation = useNavigation;
