// @bigcommerce/big-design/dist/es/components/StatefulTree/hooks/useExpandable.js
import _toConsumableArray from "@babel/runtime/helpers/esm/toConsumableArray";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { useEffect, useState } from 'react';
export var useExpandable = function useExpandable(_ref) {
  var defaultExpanded = _ref.defaultExpanded,
      onExpandedChange = _ref.onExpandedChange;

  var _useState = useState(defaultExpanded !== null && defaultExpanded !== void 0 ? defaultExpanded : []),
      _useState2 = _slicedToArray(_useState, 2),
      expandedNodes = _useState2[0],
      setExpandedNodes = _useState2[1];

  useEffect(function () {
    if (defaultExpanded) {
      setExpandedNodes(defaultExpanded);
    }
  }, [defaultExpanded]);
  useEffect(function () {
    if (typeof onExpandedChange === 'function') {
      onExpandedChange(expandedNodes);
    }
  }, [expandedNodes, onExpandedChange]);

  var onExpand = function onExpand(nodeId) {
    setExpandedNodes([].concat(_toConsumableArray(expandedNodes), [nodeId]));
  };

  var onCollapse = function onCollapse(nodeId) {
    setExpandedNodes(function (prevNodes) {
      return prevNodes.filter(function (node) {
        return node !== nodeId;
      });
    });
  };

  var onToggle = function onToggle(nodeId, isExpanded) {
    if (isExpanded) {
      onCollapse(nodeId);
    } else {
      onExpand(nodeId);
    }
  };

  return {
    expandedNodes: expandedNodes,
    onToggle: onToggle
  };
};
