// @bigcommerce/big-design/dist/cjs/components/ButtonGroup/ButtonGroup.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ButtonGroup = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _bigDesignIcons = require("@bigcommerce/big-design-icons");

var _react = _interopRequireWildcard(require("react"));

var _hooks = require("../../hooks");

var _Dropdown = require("../Dropdown");

var _Flex = require("../Flex");

var _styled = require("./styled");

var _excluded = ["iconOnly", "iconRight", "iconLeft"],
    _excluded2 = ["actions"],
    _excluded3 = ["text", "icon"];

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var excludeIconProps = function excludeIconProps(_ref) {
  var iconOnly = _ref.iconOnly,
      iconRight = _ref.iconRight,
      iconLeft = _ref.iconLeft,
      actionProps = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  return actionProps;
};

var ButtonGroup = /*#__PURE__*/(0, _react.memo)(function (_ref2) {
  var actions = _ref2.actions,
      wrapperProps = (0, _objectWithoutProperties2.default)(_ref2, _excluded2);
  var parentRef = /*#__PURE__*/(0, _react.createRef)();
  var dropdownRef = /*#__PURE__*/(0, _react.createRef)();

  var _useState = (0, _react.useState)(false),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      isMenuVisible = _useState2[0],
      setIsMenuVisible = _useState2[1];

  var _useState3 = (0, _react.useState)([]),
      _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
      actionsState = _useState4[0],
      setActionsState = _useState4[1];

  (0, _react.useEffect)(function () {
    setActionsState(actions.map(function (action) {
      return {
        isVisible: true,
        action: excludeIconProps(action),
        ref: /*#__PURE__*/(0, _react.createRef)()
      };
    }));
  }, [actions]);
  var hideOverflowedActions = (0, _react.useCallback)(function () {
    var _parentRef$current, _dropdownRef$current;

    var parentWidth = (_parentRef$current = parentRef.current) === null || _parentRef$current === void 0 ? void 0 : _parentRef$current.offsetWidth;
    var dropdownWidth = (_dropdownRef$current = dropdownRef.current) === null || _dropdownRef$current === void 0 ? void 0 : _dropdownRef$current.offsetWidth;

    if (!parentWidth || !dropdownWidth) {
      return;
    }

    var remainingWidth = parentWidth;
    var nextState = actionsState.map(function (stateObj) {
      var _stateObj$ref$current;

      var actionWidth = (_stateObj$ref$current = stateObj.ref.current) === null || _stateObj$ref$current === void 0 ? void 0 : _stateObj$ref$current.offsetWidth;

      if (!actionWidth) {
        return stateObj;
      }

      if (stateObj.action.actionType === 'destructive') {
        return _objectSpread(_objectSpread({}, stateObj), {}, {
          isVisible: false
        });
      }

      if (remainingWidth - actionWidth > dropdownWidth) {
        remainingWidth = remainingWidth - actionWidth;
        return _objectSpread(_objectSpread({}, stateObj), {}, {
          isVisible: true
        });
      }

      return _objectSpread(_objectSpread({}, stateObj), {}, {
        isVisible: false
      });
    });
    var visibleActions = actionsState.filter(function (_ref3) {
      var isVisible = _ref3.isVisible;
      return isVisible;
    });
    var nextVisibleActions = nextState.filter(function (_ref4) {
      var isVisible = _ref4.isVisible;
      return isVisible;
    });

    if (visibleActions.length !== nextVisibleActions.length) {
      setActionsState(nextState);
    }
  }, [actionsState, dropdownRef, parentRef]);
  var renderedDropdown = (0, _react.useMemo)(function () {
    return /*#__PURE__*/_react.default.createElement(_styled.StyledFlexItem, {
      "data-testid": "buttongroup-dropdown",
      isVisible: isMenuVisible,
      ref: dropdownRef,
      role: "listitem"
    }, /*#__PURE__*/_react.default.createElement(_Dropdown.Dropdown, {
      items: actionsState.filter(function (_ref5) {
        var isVisible = _ref5.isVisible;
        return !isVisible;
      }).map(function (_ref6) {
        var action = _ref6.action,
            ref = _ref6.ref;
        return {
          actionType: action.actionType,
          content: action.text,
          disabled: action.disabled,
          onItemClick: function onItemClick() {
            if (ref.current) {
              ref.current.getElementsByTagName('button')[0].click();
            }
          },
          hash: action.text.toLowerCase(),
          icon: action.icon
        };
      }),
      toggle: /*#__PURE__*/_react.default.createElement(_styled.StyledButton, {
        borderRadius: actionsState.every(function (_ref7) {
          var isVisible = _ref7.isVisible;
          return !isVisible;
        }),
        iconOnly: /*#__PURE__*/_react.default.createElement(_bigDesignIcons.MoreHorizIcon, {
          title: "more"
        }),
        variant: "secondary"
      }),
      placement: "bottom-end"
    }));
  }, [actionsState, dropdownRef, isMenuVisible]);
  var renderedActions = (0, _react.useMemo)(function () {
    return (0, _toConsumableArray2.default)(actionsState).reverse().sort(function (_ref8) {
      var isVisible = _ref8.isVisible;
      return isVisible ? -1 : 1;
    }).map(function (_ref9, key) {
      var action = _ref9.action,
          isVisible = _ref9.isVisible,
          ref = _ref9.ref;
      var text = action.text,
          icon = action.icon,
          buttonProps = (0, _objectWithoutProperties2.default)(action, _excluded3);
      return /*#__PURE__*/_react.default.createElement(_styled.StyledFlexItem, {
        "data-testid": "buttongroup-item",
        key: key,
        isVisible: isVisible,
        ref: ref,
        role: "listitem"
      }, /*#__PURE__*/_react.default.createElement(_styled.StyledButton, (0, _extends2.default)({}, buttonProps, {
        variant: "secondary"
      }), text));
    });
  }, [actionsState]);
  (0, _react.useEffect)(function () {
    var nextIsMenuVisible = actionsState.some(function (_ref10) {
      var isVisible = _ref10.isVisible;
      return !isVisible;
    });

    if (nextIsMenuVisible !== isMenuVisible) {
      setIsMenuVisible(nextIsMenuVisible);
    }
  }, [actionsState, isMenuVisible]);
  (0, _react.useEffect)(function () {
    hideOverflowedActions();
  }, [actions, parentRef, hideOverflowedActions]);
  (0, _hooks.useWindowResizeListener)(function () {
    hideOverflowedActions();
  });
  return actions.length > 0 ? /*#__PURE__*/_react.default.createElement(_Flex.Flex, (0, _extends2.default)({}, wrapperProps, {
    "data-testid": "buttongroup-wrapper",
    flexDirection: "row",
    flexWrap: "nowrap",
    ref: parentRef,
    role: "list"
  }), renderedActions, renderedDropdown) : null;
});
exports.ButtonGroup = ButtonGroup;
