// @bigcommerce/big-design/dist/es/components/Stepper/Step/styled.js
import { theme as defaultTheme } from '@bigcommerce/big-design-theme';
import { hideVisually } from 'polished';
import styled from 'styled-components';
import { Box } from '../../Box';
import { Flex } from '../../Flex';
import { Grid } from '../../Grid';
import { StyleableText } from '../../Typography/private';
export var StyledStep = /*#__PURE__*/styled(Grid).withConfig({
  displayName: "styled__StyledStep",
  componentId: "sc-kstb4o-0"
})(["flex-basis:100%;gap:", ";grid-template-areas:'light dash' 'text text';grid-template-columns:auto 1fr;", "{grid-template-columns:", " 1fr;grid-template-rows:", " 1fr;max-width:", ";min-width:", ";}"], function (_ref) {
  var theme = _ref.theme;
  return theme.spacing.xSmall;
}, function (_ref2) {
  var theme = _ref2.theme;
  return theme.breakpoints.tablet;
}, function (_ref3) {
  var theme = _ref3.theme;
  return theme.spacing.xLarge;
}, function (_ref4) {
  var theme = _ref4.theme;
  return theme.spacing.xLarge;
}, function (_ref5) {
  var theme = _ref5.theme;
  return theme.helpers.remCalc(228);
}, function (_ref6) {
  var theme = _ref6.theme;
  return theme.helpers.remCalc(96);
});
export var StyledSrOnlyText = /*#__PURE__*/styled.span.withConfig({
  displayName: "styled__StyledSrOnlyText",
  componentId: "sc-kstb4o-1"
})(["", "{", "}"], function (_ref7) {
  var theme = _ref7.theme;
  return theme.breakpoints.tablet;
}, hideVisually());
export var StyledLight = /*#__PURE__*/styled(Flex).withConfig({
  displayName: "styled__StyledLight",
  componentId: "sc-kstb4o-2"
})(["border-radius:", ";color:", ";grid-area:light;user-select:none;", "{border-radius:", ";font-size:", ";line-height:", ";}"], function (_ref8) {
  var theme = _ref8.theme;
  return theme.spacing.medium;
}, function (_ref9) {
  var theme = _ref9.theme;
  return theme.colors.white;
}, function (_ref10) {
  var theme = _ref10.theme;
  return theme.breakpoints.tablet;
}, function (_ref11) {
  var theme = _ref11.theme;
  return theme.borderRadius.circle;
}, function (_ref12) {
  var theme = _ref12.theme;
  return theme.typography.fontSize.medium;
}, function (_ref13) {
  var theme = _ref13.theme;
  return theme.lineHeight.medium;
});
export var StyledDash = /*#__PURE__*/styled(Box).withConfig({
  displayName: "styled__StyledDash",
  componentId: "sc-kstb4o-3"
})(["align-self:center;height:", ";grid-area:dash;width:100%;"], function (_ref14) {
  var theme = _ref14.theme;
  return theme.helpers.remCalc(2);
});
export var StyledText = /*#__PURE__*/styled(StyleableText).withConfig({
  displayName: "styled__StyledText",
  componentId: "sc-kstb4o-4"
})(["grid-area:text;"]);
StyledStep.defaultProps = {
  theme: defaultTheme
};
StyledSrOnlyText.defaultProps = {
  theme: defaultTheme
};
StyledLight.defaultProps = {
  theme: defaultTheme
};
StyledDash.defaultProps = {
  theme: defaultTheme
};
StyledText.defaultProps = {
  theme: defaultTheme
};
